src/Services/ChatService.php

```diff
@@ -0,0 +1,303 @@
+                 *
+                 * @var ChatParticipant $chatParticipant
+                 */
+                $chatParticipant = $chatUser->getChatParticipant($chat);
+
+                if ($chatParticipant->isNotificationOn()) {
+                    $this->dispatcher->sendNow(
+                        [$chatUser],
+                        $this->notificationsFactory->build(NotificationsType::NEW_MESSAGE)
+                    );
+                }
+                $this->participantEntityService->update($chatParticipant, [ChatParticipant::IS_READ => false]);
+                event(new MessageSentUserEvent($chatUser, $chat, $sender, $chatMessage));
+            }
+
+            return $chatMessage;
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function leaveChat(IChat $chat, IChatUser $chatUser): void
+    {
+        if (!$chat->inChat($chatUser) || $chat->isClosed()) {
+            throw new ChatException('chats.leave_error');
+        }
+
+        $this->handleTransaction(function () use ($chat, $chatUser) {
+
+            $chatParticipant = $this->participantEntityService->getRepository()->findWhere([
+                ChatParticipant::USER_ID => $chatUser->getId(),
+                ChatParticipant::CHAT_ID => $chat->getId(),
+            ]);
+
+            $this->participantEntityService->delete($chatParticipant);
+
+            if ($chat->getCreator() === $chatUser) {
+                $newCreator = null;
+                foreach ($chat->getUsers() as $user) {
+                    if ($user !== $chatUser) {
+                        $newCreator = $user;
+                        break;
+                    }
+                }
+                /**
+                 * Chat participant.
+                 *
+                 * @var Chat $chat
+                 */
+                $this->chatEntityService->update($chat, [Chat::CREATED_BY => $newCreator->getId()]);
+            }
+
+            event(new ChatLeavedEvent($chat, $chatUser));
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function markChatAsRead(IChat $chat, IChatUser $chatUser): void
+    {
+        if (!$chat->inChat($chatUser) || $chat->isClosed()) {
+            throw new ChatException('chats.leave_error');
+        }
+
+        $this->handleTransaction(function () use ($chat, $chatUser) {
+            /**
+             * Chat participant.
+             *
+             * @var ChatParticipant $chatParticipant
+             */
+            $chatParticipant = $chatUser->getChatParticipant($chat);
+            $this->participantEntityService->update($chatParticipant, [
+                ChatParticipant::IS_READ => true,
+            ]);
+        });
+    }
+
+    /**
+     * Wrap closure in db transaction.
+     *
+     * @param Closure $callback Callback which will be wrapped into transaction
+     *
+     * @return mixed
+     *
+     * @throws ChatException
+     */
+    protected function handleTransaction(Closure $callback)
```

Comment:
One of rare cases, when extracting method as trait makes sense. Probably, to one of other saritasa-laravel-common packages.